Drv/Ip/SocketComponentHelper.cpp

```diff
@@ -0,0 +1,205 @@
+        status = this->getSocketHandler().open(fd);
+        this->m_fd = fd;
+        // Call connected any time the open is successful
+        if (Drv::SOCK_SUCCESS == status) {
+            this->m_open = true;
+            this->m_lock.unlock();
+            this->connected();
+            return status;
+        }
+    }
+    this->m_lock.unlock();
+    return status;
+}
+
+bool SocketComponentHelper::isOpened() {
+    bool is_open = false;
+    this->m_lock.lock();
+    is_open = this->m_open;
+    this->m_lock.unlock();
+    return is_open;
+}
+
+SocketIpStatus SocketComponentHelper::reconnect() {
+    SocketIpStatus status = SOCK_SUCCESS;
+
+    // Handle opening
+
+    // Open a network connection if it has not already been open
+    // TODO: improve this message
+    if (this->isStarted() and (not this->isOpened())) {
+        status = this->open();
+    }
+    return status;
+}
+
+SocketIpStatus SocketComponentHelper::send(const U8* const data, const U32 size) {
+    SocketIpStatus status = SOCK_SUCCESS;
+    this->m_lock.lock();
+    NATIVE_INT_TYPE fd = this->m_fd;
+    this->m_lock.unlock();
+    // Prevent transmission before connection, or after a disconnect
+    if (fd == -1) {
+        status = this->reconnect();
+        // if reconnect wasn't successful, pass the that up to the caller
+        if(status != SOCK_SUCCESS) {
+            return status;
+        }
+        this->m_lock.lock();
+        fd = this->m_fd;
+        this->m_lock.unlock();
+    }
+    status = this->getSocketHandler().send(fd, data, size);
+    if (status == SOCK_DISCONNECTED) {
+        this->close();
+    }
+    return status;
+}
+
+void SocketComponentHelper::shutdown() {
+    this->m_lock.lock();
+    this->getSocketHandler().shutdown(this->m_fd);
+    this->m_started = false;
+    this->m_fd = -1;
+    this->m_lock.unLock();
+}
+
+void SocketComponentHelper::close() {
+    this->m_lock.lock();
+    if (this->m_fd != -1) {
+        this->getSocketHandler().close(this->m_fd);
+        this->m_fd = -1;
+    }
+    this->m_open = false;
+    this->m_lock.unlock();
+}
+
+Os::Task::Status SocketComponentHelper::join() {
+    return m_task.join();
+}
+
+void SocketComponentHelper::stop() {
+    this->m_lock.lock();
+    this->m_stop = true;
+    this->m_lock.unlock();
+    this->shutdown();  // Break out of any receives and fully shutdown
+}
+
+SocketIpStatus SocketComponentHelper::recv(U8* data, U32 &size) {
+    SocketIpStatus status = SOCK_SUCCESS;
+    // Check for previously disconnected socket
+    this->m_lock.lock();
+    NATIVE_INT_TYPE fd = this->m_fd;
+    this->m_lock.unlock();
+    if (fd == -1) {
+        return SOCK_DISCONNECTED;
+    }
+    status = this->getSocketHandler().recv(fd, data, size);
+    if (status == SOCK_DISCONNECTED) {
+        this->close();
+    }
+    return status;
+}
+
+void SocketComponentHelper::readTask(void* pointer) {
+    FW_ASSERT(pointer);
```

Comment:
Given the change in behavior of this, we should either:

1. Start before the loop
2. Assert is_started before looping

Reply:
This function is safe in that it will not break, but it won't notify the user if they failed to call start.

Reply:
Fixed at component level w/ autostart!